include/fi_mem.h

```diff
@@ -79,7 +80,7 @@ static inline void *mem_dup(const void *src, size_t size)
 		int i = (fs)->next;				\
 		assert(!freestack_isempty(fs));			\
 		(fs)->next = *((int*) &(fs)->buf[i]);		\
-		i;						\
+		&(fs)->buf[i];					\
```

Comment:
If we change the interface to push/pop pointers to entries, we should update the free list to use pointers as well.  It would mean checking that each entry is at least sizeof(void*) and changing FREESTACK_EMPTY to NULL.